tools/syz-declextract/syz-declextract.cpp

```diff
@@ -142,41 +184,88 @@ const std::string makeConst(const std::string &type, const std::string &val = "0
   if (type.empty()) {
     return "const[" + val + "]";
   }
-  if (type == "void") {
-    return "const[" + val + ", intptr]";
-  }
   return "const[" + val + ", " + type + "]";
 }
 
-const std::string getSyzType(const std::string &ctype, const bool isSyscallParam = false) {
+enum IntType {
+  INVALID_INT = 0,
+  INT_8 = 1,
+  INT_16 = 2,
+  INT_32 = 4,
+  INT_64 = 8,
+  INT_PTR,
+};
+
+IntType getIntType(const std::string &ctype, const bool isSyscallParam) {
+  // TODO: Handle arm32 passing 64bit arguments
   if (!isSyscallParam && (contains(ctype, "long long") || contains(ctype, "64"))) {
-    return contains(ctype, "be") ? "int64be" : "int64";
+    return INT_64;
   }
   if (contains(ctype, "16") || contains(ctype, "short")) {
-    return contains(ctype, "be") ? "int16be" : "int16";
+    return INT_16;
   }
-  if (contains(ctype, "8")) {
-    return contains(ctype, "be") ? "int8be" : "int8";
+  if (contains(ctype, "8") || contains(ctype, "char")) {
+    return INT_8;
   }
   if (contains(ctype, "32") || contains(ctype, "int")) {
-    return contains(ctype, "be") ? "int32be" : "int32";
+    return INT_32;
   }
-  if (contains(ctype, "void")) {
-    return "void";
+  if (contains(ctype, "void") || contains(ctype, "long")) {
+    return INT_PTR;
   }
-  if (contains(ctype, "char")) {
-    return "int8";
+  fprintf(stderr, "Unhandled int length for type: %s\n", ctype.c_str());
+  exit(1);
+}
+
+const std::string intSubtype(const std::string &name, const IntType len) {
+  switch (len) {
+  case INT_8:
+    return int8Subtype(name);
+  case INT_16:
+    return int16Subtype(name);
+  case INT_32:
+    return int32Subtype(name);
+  case INT_64:
+    return int64Subtype();
+  case INT_PTR:
+    return "intptr";
+  case INVALID_INT:
+    fprintf(stderr, "Invalid int type\n");
+    exit(1);
   }
-  return "intptr";
+}
+
+const std::string getSyzType(const std::string &ctype, const std::string &name, const bool isSyscallParam,
+                             const int bitFieldWidth = 0) {
+  IntType len = getIntType(ctype, isSyscallParam);
+  const int byteLen = len * 8;
+  if (INT_8 <= len && len <= INT_64 && contains(ctype, "be")) {
+    return "int" + std::to_string(byteLen) + "be";
+  }
+
+  std::string type;
+  if (bitFieldWidth) {
+    type = "int" + std::to_string(byteLen);
+    if (byteLen != bitFieldWidth) {
+      type += ":" + std::to_string(bitFieldWidth);
+    }
+  } else {
+    type = intSubtype(name, len);
+  }
+  if (name.empty() || contains(name, "pad") || contains(name, "unused") || contains(name, "_reserved")) {
+    return makeConst(isSyscallParam ? "" : type);
```

Comment:
I think this should be mutually exclusive with intSubtype case.

Reply:
intSubtype also returns intN types, which often match this criteria. Since intSubtype uses the name,  and none of these cases would match the cases in intSubtype(), it would just create a const[0, intN]

Reply:
"padfd", "fd_reserved" and many mode will match both cases.
Handling of "enabled" may need the same restructuring as handling of "pad".

Reply:
In these cases, what type should these fields take? The current way would make them a const. Is this wrong?

Reply:
I would say that "fd" should probably take precedence if we are unsure. But at least it shouldn't be `const[0, fd]`.